### PR TITLE
Create One_or_two library

### PR DESCRIPTION
**Pull request by enolan**
I wrote this for some refactoring of the transaction snark and snark work code, but since that work involves messing with a lot of signatures and the signatures are currently in flux I'm going to set it aside for a bit. Let's merge the library now.

### CODE REVIEW

**Comment by psteckler** [comment]
would an `of_list_exn` be useful?

File: src/lib/one_or_two/one_or_two.ml [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  module Latest = V1
+end
+
+type 'a t = 'a Stable.V1.t [@@deriving compare, equal, hash, sexp, yojson]
+
+let length = function `One _ -> 1 | `Two _ -> 2
+
+let to_list = function `One a -> [a] | `Two (a, b) -> [a; b]
+
+let group_sequence : 'a Sequence.t -> 'a t Sequence.t =
```
Comment (psteckler) [inline src/lib/one_or_two/one_or_two.ml]: I'm not of a fan of using `'` to distinguish variables, just saying

Comment (enolan) [reply src/lib/one_or_two/one_or_two.ml]: I might argue with you in another situation but once you reach the level of triple prime I agree it's bad.

**Comment by enolan** [comment]
I think `of_list_exn` would be bad. If we know it's a one or two element list already we should be using `One_or_two.t`.

File: src/lib/one_or_two/one_or_two.ml (hunk 1 above)
Comment (bkase) [reply src/lib/one_or_two/one_or_two.ml]: This is all bikeshedding but I like the `'` up to two, it's a nice way to say it's a changed version of some variable.

File: src/lib/one_or_two/intfs.ml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+type 'a t = [`One of 'a | `Two of 'a * 'a]
+
+module type Monadic = sig
```
Comment (bkase) [inline src/lib/one_or_two/intfs.ml]: Could you instead use `Monad.S` from Base? And use `Monad.Make` there instead. Then we get Let_syntax for free too.

Comment (enolan) [reply src/lib/one_or_two/intfs.ml]: huh? `One_or_two`s don't form a monad, this is the interface of monadic stuff you can do with `One_or_two`s. Like `Deferred.List.map` and friends.

Comment (bkase) [reply src/lib/one_or_two/intfs.ml]: oh I see, the `t` is from the one or two. Okay

**Pull request merged**